### ISSUE TITLE
Consider producing more friendly output with runtime drop flag checking

### ISSUE THREAD

**Issue by eefriedman** [rust-lang/rust#35101]
```
struct C;
impl Drop for C {
    fn drop(&mut self) {}
}
fn main() {
    {
        let mut c = C;
        unsafe { *(&mut c as *mut C as *mut u8) = 1; }
    }
    println!("Hello World");
}
```

This produces absolutely zero output in the playground because it generates a SIGTRAP and kills the program.  Of course, it's hard to do much at this point because the program is in an unknown, probably-corrupt state, but we could at least print a message to stderr noting what happened.

Originally reported at https://internals.rust-lang.org/t/im-moving-away-heres-my-two-cents/3767 .

**Comment by nagisa** [comment]
This is mostly solved by not having filling drop (cough… MIR… cough). In case anybody still wants to fix this issue, the relevant TRAP likely comes from code generated by https://github.com/rust-lang/rust/blob/d1df3fecdf8dc959fbd8901603a16e5bc0bfa21d/src/librustc_trans/glue.rs#L314

**Comment by nagisa** [comment]
Printing a message, though, would essentially mean extra code in every single drop glue. I wonder if that’s really the tradeoff anybody would be interested in making.

**Comment by DemiMarie** [comment]
@nagisa could this be debug-build-only?  I would imagine having the function that emits the message just write a string to stderr before raising the signal itself.

On the other hand the current implementation helps debugging because it traps into the debugger.

**Comment by arielb1** [comment]
This is not a purely zeroing drop issue: if you are using an enum instead of a struct, mangling the discriminant will cause an out-of-range switch and random UB (often a `SIGILL`).

**Comment by petevine** [comment]
Indeed, I'm getting a `SIGILL` here (debug only obviously), similar, but unrelated, to  #34427.

**Comment by DemiMarie** [comment]
Would Valgrind detect this?  If not, I suspect that Rust needs ASan and UBSan support.
